File: android/Reguerta/app/src/main/java/com/reguerta/user/data/shiftswap/ChainedShiftSwapRequestRepository.kt
package com.reguerta.user.data.shiftswap

import com.reguerta.user.domain.shifts.ShiftSwapRequest
import com.reguerta.user.domain.shifts.ShiftSwapRequestRepository

class ChainedShiftSwapRequestRepository(
    private val primary: ShiftSwapRequestRepository,
    private val fallback: ShiftSwapRequestRepository,
) : ShiftSwapRequestRepository {
    override suspend fun getAllShiftSwapRequests(): List<ShiftSwapRequest> {
        val primaryResult = primary.getAllShiftSwapRequests()
        return if (primaryResult.isNotEmpty()) primaryResult else fallback.getAllShiftSwapRequests()
    }

    override suspend fun upsertShiftSwapRequest(request: ShiftSwapRequest): ShiftSwapRequest =
        runCatching { primary.upsertShiftSwapRequest(request) }
            .getOrElse { fallback.upsertShiftSwapRequest(request) }
}

File: ios/Reguerta/Reguerta/Data/ShiftSwapRequests/ChainedShiftSwapRequestRepository.swift
import Foundation

struct ChainedShiftSwapRequestRepository<
    Primary: ShiftSwapRequestRepository,
    Fallback: ShiftSwapRequestRepository
>: ShiftSwapRequestRepository {
    let primary: Primary
    let fallback: Fallback

    func allShiftSwapRequests() async throws -> [ShiftSwapRequest] {
        try await primary.allShiftSwapRequests()
    }

    func transition(_ transition: ShiftSwapTransition) async throws -> ShiftSwapTransitionResult {
        try await primary.transition(transition)
    }
}
